src/providers/mssql/qgsmssqlgeometryparser.cpp

```diff
@@ -155,307 +201,341 @@ void QgsMssqlGeometryParser::DumpMemoryToLog( const char *pszMsg, unsigned char
+      return QgsPoint( QgsWkbTypes::PointZ, ReadY( iPoint ), ReadX( iPoint ), ReadZ( iPoint ) );
+    else if ( chProps & SP_HASMVALUES )
+      return QgsPoint( QgsWkbTypes::PointM, ReadY( iPoint ), ReadX( iPoint ), 0.0, ReadZ( iPoint ) );
+    else
+      return QgsPoint( QgsWkbTypes::Point, ReadY( iPoint ), ReadX( iPoint ) );
+  }
+  else
+  {
+    if ( ( chProps & SP_HASZVALUES ) && ( chProps & SP_HASMVALUES ) )
+      return QgsPoint( QgsWkbTypes::PointZM, ReadX( iPoint ), ReadY( iPoint ), ReadZ( iPoint ), ReadM( iPoint ) );
+    else if ( chProps & SP_HASZVALUES )
+      return QgsPoint( QgsWkbTypes::PointZ, ReadX( iPoint ), ReadY( iPoint ), ReadZ( iPoint ) );
+    else if ( chProps & SP_HASMVALUES )
+      return QgsPoint( QgsWkbTypes::PointM, ReadX( iPoint ), ReadY( iPoint ), 0.0, ReadZ( iPoint ) );
+    else
+      return QgsPoint( QgsWkbTypes::Point, ReadX( iPoint ), ReadY( iPoint ) );
   }
-  memcpy( pszWkb + nWkbLen, src, len );
-  nWkbLen += len;
 }
 
 /************************************************************************/
-/*                         CopyCoordinates()                            */
+/*                         readPointSequence()                          */
 /************************************************************************/
 
-void QgsMssqlGeometryParser::CopyCoordinates( int iPoint )
+const QgsPointSequence QgsMssqlGeometryParser::readPointSequence( int iPoint, int iNextPoint )
 {
-  if ( IsGeography )
+  if ( iPoint >= iNextPoint )
+    return QgsPointSequence();
+
+  QgsPointSequence pts;
+
+  int i = 0;
+  while ( iPoint < iNextPoint )
   {
-    CopyBytes( pszData + nPointPos + 16 * iPoint + 8, 8 ); // longitude
-    CopyBytes( pszData + nPointPos + 16 * iPoint, 8 ); // latitude
+    pts << readCoordinates( iPoint );
+    ++iPoint;
+    ++i;
   }
-  else
-    // copy geometry coords
-    CopyBytes( pszData + nPointPos + 16 * iPoint, 16 );
 
-  if ( chProps & SP_HASZVALUES )
-    CopyBytes( pszData + nPointPos + 16 * nNumPoints + 8 * iPoint, 8 ); // copy z value
+  return pts;
 }
 
 /************************************************************************/
-/*                         CopyPoint()                                  */
+/*                         readPoint()                                  */
 /************************************************************************/
 
-void QgsMssqlGeometryParser::CopyPoint( int iPoint )
+std::unique_ptr< QgsPoint > QgsMssqlGeometryParser::readPoint( int iFigure )
 {
-  // copy byte order
-  CopyBytes( &chByteOrder, 1 );
-  // copy type
-  int wkbType;
-  if ( chProps & SP_HASZVALUES )
-    wkbType = QgsWkbTypes::Point25D;
-  else
-    wkbType = QgsWkbTypes::Point;
-  CopyBytes( &wkbType, 4 );
-  // copy coordinates
-  CopyCoordinates( iPoint );
+  if ( iFigure < nNumFigures )
+  {
+    int iPoint = PointOffset( iFigure );
+    if ( iPoint < nNumPoints )
+    {
+      return qgis::make_unique< QgsPoint >( readCoordinates( iPoint ) );
```

Comment:
This may be over-optimising, but there's a number of copy operations occurring during the parsing. Potentially readCoordinates could be modified to take double x/y/z/m references and set these values in place, and then the callers could form either QgsPoint values OR unique_ptrs depending on their need.

For reference, it's also MUCH faster to create linestrings using the constructor which takes separate vectors of x/y/z/m values -- so that's a potential optimisation would could be done with a more flexible readCoordinates method.

Not a blocker, but something to keep in mind if we are aiming for peek performance for the provider.

Reply:
I've also been thinking about such change and just addressed this in aedb9e5 for the line strings. It is a bit disappointing that the circular string doesn't have such constructor, so we continue to initialize that with QgsPointSequence.

Reply:
Yes -- this is just a result of less code profiling and optimisations being implemented in the circular geometry handling, totally a consequence of linestrings being more common than curved geometries. But a very similar optimisation would also assist circular strings, and we should consider doing this sometime.